src/rust/engine/graph/src/lib.rs

```diff
@@ -653,21 +653,21 @@ impl<N: Node> Graph<N> {
           // TODO: doing cycle detection under the lock... unfortunate, but probably unavoidable
           // without a much more complicated algorithm.
           let potential_dst_id = inner.ensure_entry(dst_node.clone());
-          match Self::detect_cycle(src_id, potential_dst_id, &mut inner) {
-            Ok(true) => {
-              // Cyclic dependency: declare a dependency on a copy of the Node that is marked Cyclic.
-              inner.ensure_entry(EntryKey::Cyclic(dst_node))
-            }
-            Ok(false) => {
-              // Valid dependency.
-              trace!(
-                "Adding dependency from {:?} to {:?}",
-                inner.entry_for_id(src_id).unwrap().node(),
-                inner.entry_for_id(potential_dst_id).unwrap().node()
-              );
-              potential_dst_id
-            }
-            Err(err) => return futures::future::err(err).to_boxed(),
+          if let Some(cycle_path) = Self::detect_cycle(src_id, potential_dst_id, &mut inner) {
```

Comment:
Question outside the scope of the PR since I'm not too familiar with the code here:

Is `detect_cycle` called on every node insertion to the graph? If so, that'll represent a 
`log1 + log2 + log3 + log4 + ... + log n= log(n!) < n log(n)`  cost for constructing the graph, right?

Reply:
It is yea. But there is one heuristic that helps quite a bit: we cycle detect in the direction with the fewest edges (ie, either "up" toward roots, or "down" toward leaves). Whenever I've profiled cycle detection, it's shown up as around 3 to 5% of CPU time. I implemented a more complicated algorithm at one point as well ([here](https://github.com/twitter/pants/blob/8f9f0306e4ebb0cfc8e7aab79ebbcd1902b00b85/src/python/pants/engine/exp/scheduler.py#L92-L167)), but it didn't end up being worth it because the heuristic above seemed to be sufficient. That algorithm might be easier to implement (correctly) in rust though.